## Exploration of using Google Places API for use in mapping out science related businesses that have science related job postings in Houston, Texas. 

#### started 2017-02-17

In [104]:
import requests
import configparser
import json
import geojson

### using <a href="https://docs.python.org/3/library/configparser.html">configparser</a> to handle config file 

In [46]:
config = configparser.ConfigParser()

In [47]:
configFilePath = './config.conf'
# configparser.ConfigParser().read(configFilePath)
config.read(configFilePath)

['./config.conf']

sets key as variable that contains the google-places-api key

In [48]:
# key = configparser['google-places-api']['key1']
key = config['google-places-api']['key1']

In [49]:
key_pre = '+&key='

### test 1: Anadarko

In [50]:
search_string_place = "Anadarko+in+Houston+Texas"

In [51]:
base_url = 'https://maps.googleapis.com/maps/api/place/textsearch/json?query='

In [52]:
r = requests.get(base_url+search_string_place+key_pre+key)

In [53]:
r.text

'{\n   "html_attributions" : [],\n   "results" : [\n      {\n         "formatted_address" : "811 Main St, Houston, TX 77002, United States",\n         "geometry" : {\n            "location" : {\n               "lat" : 29.75793480000001,\n               "lng" : -95.36348679999999\n            },\n            "viewport" : {\n               "northeast" : {\n                  "lat" : 29.7593997302915,\n                  "lng" : -95.36233996970849\n               },\n               "southwest" : {\n                  "lat" : 29.7567017697085,\n                  "lng" : -95.36503793029149\n               }\n            }\n         },\n         "icon" : "https://maps.gstatic.com/mapfiles/place_api/icons/civic_building-71.png",\n         "id" : "bc48b3afcf2539828887670b23c45e5a2b58fbbb",\n         "name" : "Anadarko Petroleum Corporation",\n         "place_id" : "ChIJjdHTPDu_QIYRyt4suPDEKtU",\n         "rating" : 1,\n         "reference" : "CmRSAAAAfrLsNa2aaJxjJyxGKPU_dgwlcuYQy9pQ7y2N2scHB5h0ES

In [54]:
r.json()

{'html_attributions': [],
 'results': [{'formatted_address': '811 Main St, Houston, TX 77002, United States',
   'geometry': {'location': {'lat': 29.75793480000001,
     'lng': -95.36348679999999},
    'viewport': {'northeast': {'lat': 29.7593997302915,
      'lng': -95.36233996970849},
     'southwest': {'lat': 29.7567017697085, 'lng': -95.36503793029149}}},
   'icon': 'https://maps.gstatic.com/mapfiles/place_api/icons/civic_building-71.png',
   'id': 'bc48b3afcf2539828887670b23c45e5a2b58fbbb',
   'name': 'Anadarko Petroleum Corporation',
   'place_id': 'ChIJjdHTPDu_QIYRyt4suPDEKtU',
   'rating': 1,
   'reference': 'CmRSAAAAfrLsNa2aaJxjJyxGKPU_dgwlcuYQy9pQ7y2N2scHB5h0ESMHt-V1s2CKTseBETUvkNuER1aUtanrDgzDiDQGa6bofi1lvhrwlYIfevZ8L1JgCrxm0QnnQTLZ7GSuXqCnEhCFOi0_3yx6fztEjp5en3hNGhQXs27qyhmHSqRZ4FeXezVTKSUW1Q',
   'types': ['point_of_interest', 'establishment']},
  {'formatted_address': '1201 Lake Robbins Dr, The Woodlands, TX 77380, United States',
   'geometry': {'location': {'lat': 30.16

In [55]:
r.json()['results'][0]['geometry']['location']

{'lat': 29.75793480000001, 'lng': -95.36348679999999}

### Test 2 : Laboratories in Houston, Texas

In [56]:
search_string_place = "Laboratories+in+Houston+Texas"

In [57]:
r = requests.get(base_url+search_string_place+key_pre+key)

In [58]:
r.json()

{'html_attributions': [],
 'next_page_token': 'CoQC8wAAAJgbFIXNDC9VEWrNA_uofGfkiBjJk8DuhlYKKVwsZGFqYzAPbko6itJ-PHy37O8chuC3E6XnceTzbrp0KqaM6PeBWMASLp0tDWttL9DQ9BvECIlybDR-JAMu_8yz730FCdwD3IgQUJvpgx60kN7HJVrVdewvKtM2hylX3JQDzLsWS4ZpM68bEQMN4xHi2LLatQtowK0TjLawdJ4N6zcg1mEIDeoz-a296-pbS8Qd8rZY3d5vY51hY_YSw1Sqc8hghfcx4gCtjLn9RSvJVd52h-qPYN4kZ8oTQC-7TTvPY2wJuEQQYEFnmkaysbxv1zX8beGJ9OOZ7eiqJe86gOyVNqwSENbMe9f5qWbTu26vP_2qAAsaFDg3B8lTQlArJ6JiGXE8rbiwJUlL',
 'results': [{'formatted_address': '10401 Town Park Dr, Houston, TX 77072, United States',
   'geometry': {'location': {'lat': 29.7113274, 'lng': -95.5643411},
    'viewport': {'northeast': {'lat': 29.7129034302915,
      'lng': -95.56299041970848},
     'southwest': {'lat': 29.7102054697085, 'lng': -95.5656883802915}}},
   'icon': 'https://maps.gstatic.com/mapfiles/place_api/icons/generic_business-71.png',
   'id': '7078987cb1134e7f553da988f36b59d3f89d2644',
   'name': 'SpectraCell Laboratories, Inc.',
   'place_id': 'ChIJ_d04iyzdQIYR8ihE3

In [59]:
len(r.json()['results'])

20

In [60]:
results = r.json()['results']
for each in results:
    #print(each)
    print(each['name'])

SpectraCell Laboratories, Inc.
Houston Metallurgical Lab Inc
Access Clinical Laboratory, LLC
SGS Accutest Inc - Houston, Texas
Core Laboratories Inc
ALS Environmental
A & B Labs
Any Lab Test Now
Elite Clinical Laboratory
Texas OilTech Labs
Weatherford Labs
Polyhedron Laboratories Inc
Accu-Test Labs
Bioreference Laboratories
Clinical Pathology Labs
Clinical Pathology Labs
Clinical Pathology Labs
BM
Any Lab Test Now
Q C Laboratories Inc


### Test 2 : Oil and Gas in Houston, Texas

In [61]:
search_string_place = "Gas+Station+in+Houston+Texas"

In [62]:
r = requests.get(base_url+search_string_place+key_pre+key)

In [63]:
len(r.json()['results'])

20

In [64]:
results = r.json()['results']
for each in results:
    #print(each)
    print(each['name'])

H-E-B Fuel
Exxon
Exxon
Mobil
Exxon
Sunoco Gas Station
Exxon
Valero
Exxon
Chevron
Exxon
Exxon
Texaco Houston
Exxon
Chevron
Kroger Gas Station
Exxon
Exxon
Exxon
Valero Gas Station


### Attempting to Select a Small Dataset to Convert into geoJSON

In [66]:
# University of Houston
search_string_place__Houston = "University of Houston"

In [68]:
UnivHouston = requests.get(base_url+search_string_place__Houston+key_pre+key)

In [69]:
UnivHouston

<Response [200]>

In [78]:
UnivHouston.json()['results']

[{'formatted_address': '4800 Calhoun Rd, Houston, TX 77004, United States',
  'geometry': {'location': {'lat': 29.7199489, 'lng': -95.3422334},
   'viewport': {'northeast': {'lat': 29.7396497, 'lng': -95.325997},
    'southwest': {'lat': 29.7038121, 'lng': -95.36012260000001}}},
  'icon': 'https://maps.gstatic.com/mapfiles/place_api/icons/school-71.png',
  'id': 'f7ec825bef7fdab2793f8eff2019cde9a0c3ba13',
  'name': 'University of Houston',
  'photos': [{'height': 2988,
    'html_attributions': ['<a href="https://maps.google.com/maps/contrib/112933440860029710314/photos">Abhishek kaurav</a>'],
    'photo_reference': 'CoQBdwAAAJC95PNmruOauJ4Rmrg9iqFSa1hqq43k5SIaf4OFikcSwK2g79vvxNeeyYOSLsKWxgj9IpBdDgy3LAYTjhiQrQdbqQNI23NW5wqJSFwmDreccVkSxo3jpIqcVRixppKaJkxtoMGquggJkrtrAqj_wBf2PvOQ7eRg4uVcCywJ6jnpEhAyxkDkD1NnKug8SnNtG0oBGhRx0SXPvRYsJrsTKjec57JqLTCtFw',
    'width': 5312}],
  'place_id': 'ChIJmaG-jlq-QIYR50eCyn44nww',
  'rating': 4.4,
  'reference': 'CmRRAAAAvXQfw3vuPeSQwfmNP9fhoN2T0cKrYh35

#### if more than result, this header is included: 
{'html_attributions': [],
 'next_page_token': 'CoQC8wAAAJgbFIXNDC9VEWrNA_uofGfkiBjJk8DuhlYKKVwsZGFqYzAPbko6itJ-PHy37O8chuC3E6XnceTzbrp0KqaM6PeBWMASLp0tDWttL9DQ9BvECIlybDR-JAMu_8yz730FCdwD3IgQUJvpgx60kN7HJVrVdewvKtM2hylX3JQDzLsWS4ZpM68bEQMN4xHi2LLatQtowK0TjLawdJ4N6zcg1mEIDeoz-a296-pbS8Qd8rZY3d5vY51hY_YSw1Sqc8hghfcx4gCtjLn9RSvJVd52h-qPYN4kZ8oTQC-7TTvPY2wJuEQQYEFnmkaysbxv1zX8beGJ9OOZ7eiqJe86gOyVNqwSENbMe9f5qWbTu26vP_2qAAsaFDg3B8lTQlArJ6JiGXE8rbiwJUlL',
 'results': [{'formatted_address': '10401 Town Park Dr, Houston, TX 77072, United States',

#### otherwise the response goes straight to this: 
[{'formatted_address': '4800 Calhoun Rd, Houston, TX 77004, United States',
  'geometry': {'location': {'lat': 29.7199489, 'lng': -95.3422334},
   'viewport': {'northeast': {'lat': 29.7396497, 'lng': -95.325997},
    'southwest': {'lat': 29.7038121, 'lng': -95.36012260000001}}},
  'icon': 'https://maps.gstatic.com/mapfiles/place_api/icons/school-71.png',
  'id': 'f7ec825bef7fdab2793f8eff2019cde9a0c3ba13',
  'name': 'University of Houston',

## Brainstorming function to turn google places api result into geoJSON
#### 1. funtion that checks for whether response includes more than one result. 
#### 2. If more than one result, grabs internal ..... figure this out later.......
#### 3. assuming one result at a time, get variable that is starting of geoJSON structure

In [72]:
emptyGeoJSON = {"type":"Feature","properties":{},"geometry":{"type":"point:","coordinates":[-40.70308550521447,0.908343260182882]}}

In [73]:
emptyGeoJSON

{'geometry': {'coordinates': [-40.70308550521447, 0.908343260182882],
  'type': 'point:'},
 'properties': {},
 'type': 'Feature'}

#### Importing the functionality to make a geoJSON point using the geojson library

In [114]:
from geojson import Point, Feature, FeatureCollection
# https://github.com/frewsxcv/python-geojson

In [88]:
UnivHouston.json()['results'][0]['geometry']['location']['lat']

29.7199489

extracting information from json results into variable to add into emptyGeoJSON

In [136]:
# centerpoint for coordinates
lat = UnivHouston.json()['results'][0]['geometry']['location']['lat']
long = UnivHouston.json()['results'][0]['geometry']['location']['lng']

# properties
address = UnivHouston.json()['results'][0]['formatted_address']
id = UnivHouston.json()['results'][0]['id']
name = UnivHouston.json()['results'][0]['name']
# photos is an array
photos = UnivHouston.json()['results'][0]['photos']
place_id = UnivHouston.json()['results'][0]['place_id']
rating = UnivHouston.json()['results'][0]['rating']
reference = UnivHouston.json()['results'][0]['reference']
# types is an array
types = UnivHouston.json()['results'][0]['types']


In [138]:
lat, long

(29.7199489, -95.3422334)

In [128]:
#emptyGeoJSON[0]['geometry']['coordinates'][lat,long]
#emptyGeoJSON[0]['properties']['address':address]
emptyGeoJSON

{'geometry': {'coordinates': [-40.70308550521447, 0.908343260182882],
  'type': 'point:'},
 'properties': {},
 'type': 'Feature'}

In [142]:
type(emptyGeoJSON)

dict

##    ----------------------

In [148]:
testPoint = Point((long, lat))
testPoint

{"coordinates": [-95.3422334, 29.7199489], "type": "Point"}

In [149]:
test_geoJSOn = Feature(geometry=testPoint, properties={"name": name})
test_geoJSOn

{"geometry": {"coordinates": [-95.3422334, 29.7199489], "type": "Point"}, "properties": {"name": "University of Houston"}, "type": "Feature"}

In [150]:
test2_geoJSOn = Feature(geometry=testPoint, properties={"name": name,"address":address,"id":id,"photos":photos,"place_id":place_id,"rating":rating,"reference":reference,"types":types})
test2_geoJSOn

{"geometry": {"coordinates": [-95.3422334, 29.7199489], "type": "Point"}, "properties": {"address": "4800 Calhoun Rd, Houston, TX 77004, United States", "id": "f7ec825bef7fdab2793f8eff2019cde9a0c3ba13", "name": "University of Houston", "photos": [{"height": 2988, "html_attributions": ["<a href=\"https://maps.google.com/maps/contrib/112933440860029710314/photos\">Abhishek kaurav</a>"], "photo_reference": "CoQBdwAAAJC95PNmruOauJ4Rmrg9iqFSa1hqq43k5SIaf4OFikcSwK2g79vvxNeeyYOSLsKWxgj9IpBdDgy3LAYTjhiQrQdbqQNI23NW5wqJSFwmDreccVkSxo3jpIqcVRixppKaJkxtoMGquggJkrtrAqj_wBf2PvOQ7eRg4uVcCywJ6jnpEhAyxkDkD1NnKug8SnNtG0oBGhRx0SXPvRYsJrsTKjec57JqLTCtFw", "width": 5312}], "place_id": "ChIJmaG-jlq-QIYR50eCyn44nww", "rating": 4.4, "reference": "CmRRAAAAvXQfw3vuPeSQwfmNP9fhoN2T0cKrYh35i7uP0J1HhQglrYP8jkUwBJL8wf2cIYZjzLd110Rl29rBLxtyv5AWwA0VpZPrPiav5iImp-OhdBCpd3VF3dgOm_UCoOWoRB-1EhBIM5gMX7jGgeXYWP2Fy4e5GhTvRmKv4Gy-EJ6wC2DZikPQqUdmXQ", "types": ["university", "point_of_interest", "establishment"]}, "type": "

In [151]:
test1_FeatureCollection = FeatureCollection([test2_geoJSOn]) 
test1_FeatureCollection

{"features": [{"geometry": {"coordinates": [-95.3422334, 29.7199489], "type": "Point"}, "properties": {"address": "4800 Calhoun Rd, Houston, TX 77004, United States", "id": "f7ec825bef7fdab2793f8eff2019cde9a0c3ba13", "name": "University of Houston", "photos": [{"height": 2988, "html_attributions": ["<a href=\"https://maps.google.com/maps/contrib/112933440860029710314/photos\">Abhishek kaurav</a>"], "photo_reference": "CoQBdwAAAJC95PNmruOauJ4Rmrg9iqFSa1hqq43k5SIaf4OFikcSwK2g79vvxNeeyYOSLsKWxgj9IpBdDgy3LAYTjhiQrQdbqQNI23NW5wqJSFwmDreccVkSxo3jpIqcVRixppKaJkxtoMGquggJkrtrAqj_wBf2PvOQ7eRg4uVcCywJ6jnpEhAyxkDkD1NnKug8SnNtG0oBGhRx0SXPvRYsJrsTKjec57JqLTCtFw", "width": 5312}], "place_id": "ChIJmaG-jlq-QIYR50eCyn44nww", "rating": 4.4, "reference": "CmRRAAAAvXQfw3vuPeSQwfmNP9fhoN2T0cKrYh35i7uP0J1HhQglrYP8jkUwBJL8wf2cIYZjzLd110Rl29rBLxtyv5AWwA0VpZPrPiav5iImp-OhdBCpd3VF3dgOm_UCoOWoRB-1EhBIM5gMX7jGgeXYWP2Fy4e5GhTvRmKv4Gy-EJ6wC2DZikPQqUdmXQ", "types": ["university", "point_of_interest", "establishment

In [159]:
with open('../DATA/data5.geojson', 'w') as f:
     json.dump(test1_FeatureCollection, f)

## Test 3 - BP

In [165]:
# BP
search_string_place__BP = "BP+Houston"

In [166]:
BP_results = requests.get(base_url+search_string_place__BP+key_pre+key)

In [171]:
BP_results.json()['results']

[{'formatted_address': '201 Helios Way, Houston, TX 77079, United States',
  'geometry': {'location': {'lat': 29.7827932, 'lng': -95.6355662},
   'viewport': {'northeast': {'lat': 29.7841507302915,
     'lng': -95.6347346697085},
    'southwest': {'lat': 29.7814527697085, 'lng': -95.6374326302915}}},
  'icon': 'https://maps.gstatic.com/mapfiles/place_api/icons/generic_business-71.png',
  'id': 'e46f6a874424cc053595fa14766d20e09d95c787',
  'name': 'BP Energy Company',
  'place_id': 'ChIJg--izefbQIYRtgvWPBmGz80',
  'rating': 5,
  'reference': 'CmRSAAAAQ5ra2hUoATyQBD-Q2DRAAz66t8CG_G52GRhdHfRZWon-X0SbMxcjmFG0myG6Ntc1uA4zHveGJOrnE2gFHox56VdwOsyDKxeYLm9qKp_F6_UquqW20n-Iv3rW4a7xLFCYEhACXZLXPawOQB18kXJCvXn8GhToRJrdM9HHstm5ZSC7VFHQD5V2-Q',
  'types': ['point_of_interest', 'establishment']},
 {'formatted_address': '501 Westlake Park Blvd, Houston, TX 77079, United States',
  'geometry': {'location': {'lat': 29.7810751, 'lng': -95.62888369999999},
   'viewport': {'northeast': {'lat': 29.782079680

In [169]:
len(BP_results.json()['results'])

8

In [200]:
# function that takes a datset that is 
def makeFeatureCollectionsFromPlaces(dataset_I,nameOfSearch):
    dataset = dataset_I.json()['results']
    array_of_features = []
    for each in BP_result:
        # centerpoint for coordinates
        print(each)
        lat = each['geometry']['location']['lat']
        long = each['geometry']['location']['lng']
        # properties
        address = each['formatted_address']
        id = each['id']
        name = each['name']
        #try:
        #    photos = each['photos']
        #except AttributeError:
        #    print("photos , attribute error but kept going")
        # photos is an array
        #photos = each['photos']
        place_id = each['place_id']
        #try:
        #    rating = each['rating']
        #except AttributeError:
        #    print("rating , attribute error but kept going")
        reference = each['reference']
        # types is an array
        types = each['types']
        # 
        testPoint = Point((long, lat))
        test2_geoJSOn = Feature(geometry=testPoint, properties={"name": name,"address":address,"id":id,"photos":photos,"place_id":place_id,"rating":rating,"reference":reference,"types":types})
        array_of_features.extend([test2_geoJSOn])
    new_FeatureCollection = FeatureCollection(array_of_features) 
    with open('../DATA/'+nameOfSearch+'_test.geojson', 'w') as f:
        json.dump(new_FeatureCollection, f)


In [201]:

BP_results_geojson = makeFeatureCollectionsFromPlaces(BP_results,"BP_results")

{'types': ['point_of_interest', 'establishment'], 'id': 'e46f6a874424cc053595fa14766d20e09d95c787', 'icon': 'https://maps.gstatic.com/mapfiles/place_api/icons/generic_business-71.png', 'geometry': {'viewport': {'northeast': {'lng': -95.6347346697085, 'lat': 29.7841507302915}, 'southwest': {'lng': -95.6374326302915, 'lat': 29.7814527697085}}, 'location': {'lng': -95.6355662, 'lat': 29.7827932}}, 'formatted_address': '201 Helios Way, Houston, TX 77079, United States', 'name': 'BP Energy Company', 'reference': 'CmRSAAAAQ5ra2hUoATyQBD-Q2DRAAz66t8CG_G52GRhdHfRZWon-X0SbMxcjmFG0myG6Ntc1uA4zHveGJOrnE2gFHox56VdwOsyDKxeYLm9qKp_F6_UquqW20n-Iv3rW4a7xLFCYEhACXZLXPawOQB18kXJCvXn8GhToRJrdM9HHstm5ZSC7VFHQD5V2-Q', 'place_id': 'ChIJg--izefbQIYRtgvWPBmGz80', 'rating': 5}
{'types': ['point_of_interest', 'establishment'], 'photos': [{'height': 2988, 'photo_reference': 'CoQBdwAAANBQrgK3sXZBt5konU2kf1T0pP-SvApM55TqMowhaX-hN-5_SIuZa0RqmqnPMWXeN-StHXTxOiFr6El8aSyhU4BwpP5ob5_VwnjLuC_Jcu9-uADHCuAAdtRCvXPj4lEEtqg

In [179]:
dataset = BP_results
dataset = dataset.json()['results']
dataset

[{'formatted_address': '201 Helios Way, Houston, TX 77079, United States',
  'geometry': {'location': {'lat': 29.7827932, 'lng': -95.6355662},
   'viewport': {'northeast': {'lat': 29.7841507302915,
     'lng': -95.6347346697085},
    'southwest': {'lat': 29.7814527697085, 'lng': -95.6374326302915}}},
  'icon': 'https://maps.gstatic.com/mapfiles/place_api/icons/generic_business-71.png',
  'id': 'e46f6a874424cc053595fa14766d20e09d95c787',
  'name': 'BP Energy Company',
  'place_id': 'ChIJg--izefbQIYRtgvWPBmGz80',
  'rating': 5,
  'reference': 'CmRSAAAAQ5ra2hUoATyQBD-Q2DRAAz66t8CG_G52GRhdHfRZWon-X0SbMxcjmFG0myG6Ntc1uA4zHveGJOrnE2gFHox56VdwOsyDKxeYLm9qKp_F6_UquqW20n-Iv3rW4a7xLFCYEhACXZLXPawOQB18kXJCvXn8GhToRJrdM9HHstm5ZSC7VFHQD5V2-Q',
  'types': ['point_of_interest', 'establishment']},
 {'formatted_address': '501 Westlake Park Blvd, Houston, TX 77079, United States',
  'geometry': {'location': {'lat': 29.7810751, 'lng': -95.62888369999999},
   'viewport': {'northeast': {'lat': 29.782079680

In [183]:
dataset2 = dataset[2]['geometry']['location']['lat']
dataset2

29.728147